### PR TITLE
Gate: disallow .status.loadBalancer on non-LB svc

### LINKED ISSUE #119700: Service status.loadBalancer.ingress can be set when type != LoadBalancer
xref 119454

I think this may be one of the few cases where we can actually change the behavior in a "breaking" way.

User creates service with `.type=ClusterIP`, `.status.loadBalancer` is empty
Controller or user updates the status subresource, sets `.spec.loadBalancer.ingress[].ip` (not wiped or validated, should be)
User updates `service spec` in any way, `.staus.loadBalancer` is wiped

It's clear that the `.status.loadBalancer` is not meant to be set when `.type` is not "LoadBalancer".

The question is:

1) Fail validation?
2) Silently discard the value?
3) Warn and then discard the value?
4) Just warn and leave it.

Because we allowed it in the past, but it is so volatile, I lean towards 3.

@aojea and @khenidak - the people most familair with this type.

Edit: PrepareForUpdate() runs before warnings.  So:

(1) is easy
(2) is easy
(4) is slightly less easy than (2)
(3) is slightly less easy than (4)

### PR DESCRIPTION
**Pull request by thockin**
Fixes #119700

The fact that the .status.loadBalancer field can be set while .spec.type is not "LoadBalancer" is a flub.  Any spec update will already clear .status.ingress, so it's hard to really rely on this.  After this change, updates which try to set this combination will fail validation.

Existing cases of this will not be broken.  Any spec/metadata update will clear it (no error) and this is the only stanza of status.

New gate "AllowServiceLBStatusOnNonLB" is off by default, but can be enabled if this change actually breaks someone, which seems exceeedingly unlikely.

/kind bug

```release-note
Setting the `status.loadBalancer` of a Service whose `spec.type` is not `"LoadBalancer"` was previously allowed, but any update to the `metadata` or `spec` would wipe that field. Setting this field is no longer permitted unless `spec.type` is  `"LoadBalancer"`.  In the very unlikely event that this has unexpected impact, you can enable the `AllowServiceLBStatusOnNonLB` feature gate, which will restore the previous behavior.  If you do need to set this, please file an issue with the Kubernetes project to help contributors understand why you need it.
```

### CODE REVIEW

**Comment by thockin** [comment]
I'd like to make an argument for this to go into 1.28, but if it can't make the .0 then we should wait until 29 (and fix comments in this commit).

@kubernetes/release-managers

**Comment by pacoxu** [comment]
Unit test failed in TestServiceStatusStrategy for this change,
Other failure can be fixed by https://github.com/kubernetes/kubernetes/pull/119454.

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Setting the `status.loadBalancer` of a Service whose `spec.type` is not "LoadBalancer" was previously allowed, but any update to the `metadata` or `spec` would wipe that field.  Setting this field is no longer permitted unless `spec.type` is  "LoadBalancer".  In the very unlikely event that this has unexpected impact, the "AllowServiceLBStatusOnNonLB" feature gate may be set to true, which will restore the previous behavior.  If you need to set this, please file an issue with the Kubernetes project to help us understand why.
+Setting the `status.loadBalancer` of a Service whose `spec.type` is not `"LoadBalancer"` was previously allowed, but any update to the `metadata` or `spec` would wipe that field.
+Setting this field is no longer permitted unless `spec.type` is  `"LoadBalancer"`.  In the very unlikely event that this has unexpected impact, you can enable the `AllowServiceLBStatusOnNonLB` feature gate, which will restore the previous behavior.
+If you need to set this, please file an issue with the Kubernetes project to help contributors understand why you need it.
```

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -5441,7 +5441,7 @@ func ValidateServiceUpdate(service, oldService *core.Service) field.ErrorList {
 // ValidateServiceStatusUpdate tests if required fields in the Service are set when updating status.
 func ValidateServiceStatusUpdate(service, oldService *core.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
-	allErrs = append(allErrs, ValidateLoadBalancerStatus(&service.Status.LoadBalancer, field.NewPath("status", "loadBalancer"))...)
+	allErrs = append(allErrs, ValidateLoadBalancerStatus(&service.Status.LoadBalancer, field.NewPath("status", "loadBalancer"), &service.Spec)...)
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:5444]: why not check directly here?

```suggestion
if !utilfeature.DefaultFeatureGate.Enabled(features.AllowServiceLBStatusOnNonLB) && spec.Type != core.ServiceTypeLoadBalancer && len(status.Ingress) != 0 {
		allErrs = append(allErrs, field.Forbidden(ingrPath, "may only be used when `spec.type` is 'LoadBalancer'"))
		} else {
	allErrs = append(allErrs, ValidateLoadBalancerStatus(&service.Status.LoadBalancer, field.NewPath("status", "loadBalancer"), &service.Spec)...)
	}
```

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -44,6 +44,13 @@ const (
 	// Enable usage of Provision of PVCs from snapshots in other namespaces
 	CrossNamespaceVolumeDataSource featuregate.Feature = "CrossNamespaceVolumeDataSource"
 
+	// owner: @thockin
+	// deprecated: v1.28
```
Comment (aojea) [inline pkg/features/kube_features.go:48]:
```suggestion
	// deprecated: v1.29
```

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go:5444]: it felt more contained in the status-specific path.  Other than avoiding one argument, is there a reason I am not seeing?

**Comment by thockin** [comment]
Rebased.  Unit test updated.

**Comment by thockin** [comment]
I fixed the integration test, I think.  The rest need the e2e fix.

**Comment by aojea** [comment]
>  failed [FAILED] Could not patch service status%!(EXTRA *errors.StatusError=Service "test-service-7pblz" is invalid: status.loadBalancer.ingress: Forbidden: may only be used when `spec.type` is 'LoadBalancer'): Service "test-service-7pblz" is invalid: status.loadBalancer.ingress: Forbidden: may only be used when `spec.type` is 'LoadBalancer'
In [It] at: test/e2e/network/service.go:3406 @ 08/18/23 23:19:51.638

proof this fixes the problem

/retest

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (aojea) [reply pkg/apis/core/validation/validation.go:5444]: I was trying to avoid to plumb the spec on the status path

File: test/integration/etcd/data.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 			ExpectedEtcdPath: "/registry/configmaps/" + namespace + "/cm1",
 		},
 		gvr("", "v1", "services"): {
-			Stub:             `{"metadata": {"name": "service1"}, "spec": {"externalName": "service1name", "ports": [{"port": 10000, "targetPort": 11000}], "selector": {"test": "data"}}}`,
+			Stub:             `{"metadata": {"name": "service1"}, "spec": {"type": "LoadBalancer", "ports": [{"port": 10000, "targetPort": 11000}], "selector": {"test": "data"}}}`,
```
Comment (aojea) [inline test/integration/etcd/data.go:48]: why this change, why do you have WIP in the commit?

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (aojea) [reply pkg/apis/core/validation/validation.go:5444]: my point is that if we make `service.Spec` an argument of `ValidateLoadBalancerStatus` we may open a door to all kind of validations there , in this case have `ValidateLoadBalancerStatus` that is can be called from any `serviceType`, I think that this is our problem, we allow it from any service type.

with my suggesting once we remove the feature gate the code seems easy to read, as `ValidateLoadBalancerStatus` only will validate the status for types LoadBalancer that is what you expect from the name

File: test/integration/etcd/data.go (hunk 3 above)
Comment (thockin) [reply test/integration/etcd/data.go:48]: There are a couple tests which interate all of the API types looking for `/status` subresources.  There's a table of inputs for spec and another table for status, which it tries to `apply` to `/status`.  This test tried to apply `status.loadBalancer.ingress` while the spec is `type: ClusterIP`.  That fails now.

"WIP" in comment because I was not 100% confident in it.

**Comment by thockin** [comment]
I don't understand the remaining integration fail so I am rebasing and will see if it pops up again.

**Comment by thockin** [comment]
I think I got it now.  These tests are WAY too brittle.

**Comment by thockin** [comment]
is green now!

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -7053,32 +7053,37 @@ var (
...
 			}
-			if isIP := (netutils.ParseIPSloppy(ingress.Hostname) != nil); isIP {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("hostname"), ingress.Hostname, "must be a DNS name, not an IP address"))
+
+			if len(ingress.Hostname) > 0 {
+				for _, msg := range validation.IsDNS1123Subdomain(ingress.Hostname) {
+					allErrs = append(allErrs, field.Invalid(idxPath.Child("hostname"), ingress.Hostname, msg))
+				}
+				if isIP := (netutils.ParseIPSloppy(ingress.Hostname) != nil); isIP {
+					allErrs = append(allErrs, field.Invalid(idxPath.Child("hostname"), ingress.Hostname, "must be a DNS name, not an IP address"))
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:7085]: tehcnically I think an IPv4 is a valid dns name, no?

Comment (aojea) [reply pkg/apis/core/validation/validation.go:7085]: git blame says this was this way since forever so 🤷

**Comment by aojea** [comment]
3 greens on the last runs of integration job

https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=119789

/lgtm

File: pkg/apis/core/validation/validation.go (hunk 4 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go:7085]: I think DNS says a hostname which is all numeric is not valid, precisely because it is ambiguous.

Comment (aojea) [reply pkg/apis/core/validation/validation.go:7085]: I do not why I ask this things, now I had to spend some time going through the RFCs just to know the truth 😄 

I found this in https://www.ietf.org/rfc/rfc1123.txt

```
2.  GENERAL ISSUES

   This section contains general requirements that may be applicable to
   all application-layer protocols.

   2.1  Host Names and Numbers

      The syntax of a legal Internet host name was specified in RFC-952
      [DNS:4].  One aspect of host name syntax is hereby changed: the
      restriction on the first character is relaxed to allow either a
      letter or a digit.  Host software MUST support this more liberal
      syntax.

      Host software MUST handle host names of up to 63 characters and
      SHOULD handle host names of up to 255 characters.

      Whenever a user inputs the identity of an Internet host, it SHOULD
      be possible to enter either (1) a host domain name or (2) an IP
      address in dotted-decimal ("#.#.#.#") form.  The host SHOULD check
      the string syntactically for a dotted-decimal number before
      looking it up in the Domain Name System.

      DISCUSSION:
           This last requirement is not intended to specify the complete
           syntactic form for entering a dotted-decimal host number;
           that is considered to be a user-interface issue.  For
           example, a dotted-decimal number must be enclosed within
           "[ ]" brackets for SMTP mail (see Section 5.2.17).  This
           notation could be made universal within a host system,
           simplifying the syntactic checking for a dotted-decimal
           number.

           If a dotted-decimal number can be entered without such
           identifying delimiters, then a full syntactic check must be
           made, because a segment of a host domain name is now allowed
           to begin with a digit and could legally be entirely numeric
           (see Section 6.1.2.4).  However, a valid host name can never
           have the dotted-decimal form #.#.#.#, since at least the
           highest-level component label will be alphabetic.
```
